scripts/cot/botania/flowers/genereting/campanimia.zs

```diff
@@ -0,0 +1,168 @@
+campanimia.onUpdate = function(subtile, world, pos) { 
+    if(world.isRemote()) return;
+    if(world.time%20 != 7) return; 
+    val cruciblesPos = getCruciblesPos(world, pos);
+    if(cruciblesPos.length == 0) return;
+    val crucibles = getCrucibles(world, cruciblesPos);
+    val aspects = getAspectsList(crucibles);
+    if(aspects.length == 0) return;
+    val generatedMana = pow(10,pow(aspects.length,0.475) - 1) as int;
+    if(subtile.getMaxMana() - subtile.getMana() < generatedMana) return;
+    drinkEssentia(world, cruciblesPos);
+    subtile.addMana(generatedMana);
+    playSound("thaumcraft:spill", pos, world);
+    return;
+};
+campanimia.register();
+
+function getCruciblesPos(world as IWorld, posFlower as IBlockPos) as IBlockPos[] {
+    var cruciblesPos = [] as IBlockPos[];
+    val poss = [
+        toBlockPos(posFlower.x + 1, posFlower.y, posFlower.z),
+        toBlockPos(posFlower.x - 1, posFlower.y, posFlower.z),
+        toBlockPos(posFlower.x, posFlower.y, posFlower.z + 1),
+        toBlockPos(posFlower.x, posFlower.y, posFlower.z - 1),
+    ] as IBlockPos[];
+
+    for pos in poss {
+        val block = world.getBlock(pos);
+        if(isNull(block)
+        || isNull(block.data)
+        || isNull(block.data.id)
+        || block.data.id!="thaumcraft:tilecrucible"
+        || block.data.Amount<100
+        || block.data.Heat!=200
+        || block.data.Aspects.length==0) continue;
+        cruciblesPos += pos;
+    }
+
+    return cruciblesPos;
+}
+
+function toBlockPos(x as int, y  as int, z  as int) as IBlockPos{
+    return crafttweaker.util.Position3f.create(x, y, z) as IBlockPos;
+}
+
+function getCrucibles(world as IWorld, cruciblesPosList as IBlockPos[]) as IBlock[]{
+    var crucibles = [] as IBlock[];
+    for cruciblePos in cruciblesPosList{
+        crucibles += world.getBlock(cruciblePos);
+    }
+    return crucibles;
+}
+
+function getAspectsList(crucibles as IBlock[]) as IData { 
+    var aspects as IData = [];
+    for crucible in crucibles {
+        for i in 0 .. crucible.data.Aspects.length{
+            aspects += [crucible.data.Aspects[i].key];
+        }
+    }
+    return aspects;
+}
+
+function drinkEssentia(world as IWorld, cruciblesPosList as IBlockPos[]) as void {
+    for cruciblePos in cruciblesPosList {
+        var overwriteData = [] as IData;
+        val crucible = world.getBlock(cruciblePos);
+        for i in 0 .. crucible.data.Aspects.length {
+            if(crucible.data.Aspects[i].amount>1) overwriteData = overwriteData.deepUpdate([{amount: (crucible.data.Aspects[i].amount - 1), key: crucible.data.Aspects[i].key}], MERGE);
+        }
+        val newData = crucible.data.deepUpdate({Aspects: overwriteData, Amount: (crucible.data.Amount - 100)}, {Aspects: OVERWRITE, Amount: OVERWRITE});
+        world.setBlockState(world.getBlockState(cruciblePos), newData, cruciblePos);
+    }
+    return;
+}
+
+function playSound(str as string, pos as IBlockPos, world as IWorld) as void{
```

Comment:
This function is plain copy-paste of other `playSound` functions in other file?
This may be refactored. Please gather all `playSound` calls in one place, something like `scripts/lib/sound.zs` and call it by script reference. Dont forget to add `loader contenttweaker`.

Advice: if your file called `sount.zs` would be good to call function not `playSound()` but just `play()` since we already have `sound` in call path `scripts.lib.sound.play()`.